File: apps/api/src/common/http/pagination.ts
import type { CursorMeta, PageMeta, PaginatedResponse } from '@gamescore/types';

import { BadRequestError } from '../errors/app.exception';
import { ERROR_CODES } from '../errors/error-codes';

export const DEFAULT_PAGE_SIZE = 20;
export const MAX_PAGE_SIZE = 50;
export const DEFAULT_CURSOR_SIZE = 20;
export const MAX_CURSOR_SIZE = 50;

export function clampPage(page?: number): number {
  if (!page || !Number.isFinite(page)) return 1;
  return Math.max(1, Math.trunc(page));
}

export function clampLimit(limit?: number, fallback = DEFAULT_PAGE_SIZE): number {
  if (!limit || !Number.isFinite(limit)) return fallback;
  return Math.min(MAX_PAGE_SIZE, Math.max(1, Math.trunc(limit)));
}

export function pageMeta(total: number, page: number, limit: number): PageMeta {
  const totalPages = total === 0 ? 0 : Math.ceil(total / limit);
  return {
    page,
    limit,
    total,
    totalPages,
    hasNextPage: page < totalPages,
  };
}

export function paginated<T>(items: T[], total: number, page: number, limit: number): PaginatedResponse<T> {
  return { items, meta: pageMeta(total, page, limit) };
}

export function encodeCursor(payload: unknown): string {
  return Buffer.from(JSON.stringify(payload), 'utf8').toString('base64url');
}

export function decodeCursor<T>(cursor: string): T {
  try {
    const parsed: unknown = JSON.parse(Buffer.from(cursor, 'base64url').toString('utf8'));
    return parsed as T;
  } catch {
    throw new BadRequestError(ERROR_CODES.INVALID_CURSOR, 'Invalid pagination cursor');
  }
}

export function cursorMeta(limit: number, nextCursor: string | null): CursorMeta {
  return {
    limit,
    nextCursor,
    hasNextPage: nextCursor !== null,
  };
}

File: apps/api/src/GameScore.Application/Http/Pagination.cs
using System.Text;
using System.Text.Json;
using GameScore.Application.Errors;

namespace GameScore.Application.Http;

public static class Pagination
{
    public const int DefaultPageSize = 20;
    public const int MaxPageSize = 50;
    public const int DefaultCursorSize = 20;
    public const int MaxCursorSize = 50;

    public static int ClampPage(int? page) =>
        page is null or <= 0 ? 1 : page.Value;

    public static int ClampLimit(int? limit, int fallback = DefaultPageSize)
    {
        if (limit is null or <= 0)
        {
            return fallback;
        }

        return Math.Min(MaxPageSize, Math.Max(1, limit.Value));
    }

    public static object PageMeta(int total, int page, int limit)
    {
        var totalPages = total == 0 ? 0 : (int)Math.Ceiling(total / (double)limit);
        return new
        {
            page,
            limit,
            total,
            totalPages,
            hasNextPage = page < totalPages,
        };
    }

    public static object Paginated<T>(IReadOnlyList<T> items, int total, int page, int limit) =>
        new { items, meta = PageMeta(total, page, limit) };

    public static string EncodeCursor(object payload) =>
        Convert.ToBase64String(Encoding.UTF8.GetBytes(JsonSerializer.Serialize(payload)))
            .TrimEnd('=')
            .Replace('+', '-')
            .Replace('/', '_');

    public static T DecodeCursor<T>(string cursor)
    {
        try
        {
            var padded = cursor.Replace('-', '+').Replace('_', '/');
            switch (padded.Length % 4)
            {
                case 2: padded += "=="; break;
                case 3: padded += "="; break;
            }

            var json = Encoding.UTF8.GetString(Convert.FromBase64String(padded));
            return JsonSerializer.Deserialize<T>(json)!;
        }
        catch
        {
            throw new BadRequestException(ErrorCodes.InvalidCursor, "Invalid pagination cursor");
        }
    }

    public static object CursorMeta(int limit, string? nextCursor) =>
        new
        {
            limit,
            nextCursor,
            hasNextPage = nextCursor is not null,
        };
}
